src/handlers/pug_handler.ts

```diff
@@ -74,3 +67,86 @@ export async function friends(req: any, res: express.Response) {
 
   res.render('friends', context);
 }
+
+export async function activities(req: any, res: express.Response) {
+  const log = getLogger('activities', req);
+  const context = getSharedTemplateContext(req, log);
+
+  if (!isAuthenticated(req, res, log)) {
+    return;
+  }
+
+  await syncStravaActivities(req, res, log);
+
+  context.activities = await req.context.daos.activity.findActivitiesByUser(
+    req.session.user.id,
+  );
+
+  res.render('activities', context);
+}
+async function syncStravaActivities(
+  req: any,
+  res: express.Response,
+  log: LoggerInstance,
+) {
+  const stravaProvider = new StravaProviderDAO(
+    req.session.user.id,
+    req.session.user.accessToken,
+    req.context.loggerFactory,
+  );
+
+  // get the last time activities were synced for the user. if activites have never been synced, get the activities for
+  // the past two weeks.
+  let user = await req.context.daos.user.findById(req.session.user.id);
+  const nextActivityDate = !isNullOrUndefined(user.lastActivitiesSyncedAt)
+    ? new Date(new Date(user.lastActivitiesSyncedAt).getTime())
+    : new Date(new Date().getTime() - 1209600 * 1000); // TODO: make this configurable. Currently set to past two weeks.
+
+  // get all activities from strava after the time activities were last synced.
+  const stravaActivities = await stravaProvider.getActivities(nextActivityDate);
+  _.map(stravaActivities, stravaActivity => {
+    stravaActivity.userId = req.session.user.id;
+    return stravaActivity;
+  });
+
+  let newLastSyncedAt = new Date();
+
+  // persist the activities
+  await Promise.all(
+    _.map(stravaActivities, activity => {
+      return req.context.daos.activity.create(activity);
```

Comment:
Is it possible to make this a bulk request against mongo for performance?

Reply:
Definitely.